ui/src/storage/storage.tsx

```diff
@@ -0,0 +1,81 @@
+import Loading from "components/loading";
+import { useAppDispatch } from "hooks";
+import { useCallback } from "react";
+import { useAsync } from "react-async";
+import { AnyAction, Dispatch, Middleware, MiddlewareAPI } from "redux";
+import { loadUserData, RootState } from "rootReducer";
+import * as tanagra from "tanagra-api";
+
+const currentVersion = 1;
+
+export interface StoragePlugin {
+  store(data: tanagra.UserData): Promise<void>;
+  load(): Promise<tanagra.UserData | undefined>;
+}
+
+export const storeUserData: Middleware<unknown, RootState> =
+  (store: MiddlewareAPI<Dispatch<AnyAction>>) =>
+  (next: Dispatch<AnyAction>) =>
+  (action: AnyAction): AnyAction => {
+    const result = next(action);
+    if (storagePlugin) {
+      const state = store.getState();
+
+      const data: tanagra.UserData = {
+        version: currentVersion,
+        cohorts: state.cohorts,
+        conceptSets: state.conceptSets,
+      };
+
+      storagePlugin.store(data);
+    }
+    return result;
+  };
+
+export function LoadingUserData(props: { children?: React.ReactNode }) {
+  const dispatch = useAppDispatch();
+  const status = useAsync<tanagra.UserData>(
+    useCallback(() => {
+      if (!storagePlugin) {
+        throw new Error("No storage plugin configured.");
+      }
+
+      return storagePlugin.load().then((data?: tanagra.UserData) => {
```

Comment:
In the event that this fails, it seems like ideally we would show an error message, and then alow the user to proceed with no stored artifacts; perhaps also disabling localstorage in general so we're not overwriting their data.

If we don't have error message handling like this, next best is just log the error and allow the user to continue.